seed/seedwriter/writer.go

```diff
@@ -665,18 +734,67 @@ func (w *Writer) SetInfo(sn *SeedSnap, info *snap.Info) error {
 
 	if sn.local {
 		sn.SnapRef = info
-		// nothing more to do
-		return nil
+		return w.assignLocalComponents(sn)
 	}
 
 	p, err := w.tree.snapPath(sn)
 	if err != nil {
 		return err
 	}
 	sn.Path = p
+
+	return nil
+}
+
+type ByCompName []SeedComponent
+
+func (c ByCompName) Len() int           { return len(c) }
+func (c ByCompName) Less(i, j int) bool { return c[i].ComponentName < c[j].ComponentName }
+func (c ByCompName) Swap(i, j int)      { c[i], c[j] = c[j], c[i] }
+
+func (w *Writer) assignLocalComponents(sn *SeedSnap) error {
+
+	// Add matching local components now that we know the snap name
+	for path, compInfo := range w.byPathLocalComps {
+		if compInfo.Component.SnapName != sn.SnapName() {
+			continue
+		}
+
+		// Check if the component is defined by the snap
+		compInSnap, ok := sn.Info.Components[compInfo.Component.ComponentName]
+		if !ok {
+			return fmt.Errorf("component %s is not defined by snap %s",
+				compInfo.Component.ComponentName, sn.SnapName())
+		}
+		// and if types match
+		if compInSnap.Type != compInfo.Type {
+			return fmt.Errorf("component %s has type %s while snap %s defines type %s for it",
+				compInfo.Component.ComponentName, compInfo.Type,
+				sn.SnapName(), compInSnap.Type)
+		}
+
+		// now we can add to the snap
+		sn.Components = append(sn.Components, SeedComponent{
+			ComponentRef: compInfo.Component,
+			Path:         path,
+			Info:         compInfo,
+		})
+		// Remove so we can later check if some component path was unused
+		delete(w.byPathLocalComps, path)
```

Comment:
do we need to check somewhere whether we exhausted the given paths? also do we need to check somewhere that if we have fully local components but the grade is not dangerous we need to give the usual error for those situations?

Reply:
also now that the data is attached to the Write, tracking by deleteing is a bit odd, maybe we should track what was used/seen differently

Reply:
The check was happening in `image_linux.go` (see call to `s.w.ByPathLocalComponents()`). I have changed this though so now we do not remove local components for the map and instead it checks if there are components whose owner is not amongst the local snaps.

Regarding the grade, we admit local components only if local snaps are in the options, so the error is still shown.